### PR TITLE
HC_SR04 Distance Sensor Driver

### PR DESCRIPTION
**Pull request by cristianaprecup**
### Pull Request Overview

This pull request adds HIL and a system call driver for distance and a driver for HC-SR04 Distance Sensor.

The files added for this capsules are:
- `capsules/extra/src/distance.rs`
- `capsules/extra/src/hc_sr04.rs`

### Testing Strategy

This pull request was tested on a Raspberry Pi Pico W and a HC-SR04 distance sensor.

### TODO or Help Wanted

This pull request needs feedback.

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

File: capsules/extra/src/hc_sr04.rs
Comment (brghena) [inline capsules/extra/src/hc_sr04.rs:18]: Something doesn't math right here.

`343 m/s * 0.05 seconds = 17.15 meters` (38 ms -> 13 meters)

That's probably halved, since you need to transmit and receive the echo, but still doesn't work out to 4 meters...

File: capsules/extra/src/hc_sr04.rs
Comment (brghena) [inline capsules/extra/src/hc_sr04.rs:79]: Add a newline between the end of a function and the start of the docs for the next function, please

File: capsules/extra/src/hc_sr04.rs
Comment (brghena) [inline capsules/extra/src/hc_sr04.rs:84]: Where does the 10 ms value here come from?

Oh, I realized it's the trigger pulse duration. The datasheet notes that as 10 us, not 10 ms. You could reduce this quite a bit (1 millisecond is still probably fine?).

File: capsules/extra/src/hc_sr04.rs
Comment (brghena) [inline capsules/extra/src/hc_sr04.rs:115]: Should really do these three in reverse order. That way if an edge won't occur until you've enabled interrupts, and you won't get an interrupt until you've changed `status`.

File: capsules/extra/src/hc_sr04.rs
Comment (brghena) [inline capsules/extra/src/hc_sr04.rs:130]: Similarly reorder these

File: capsules/extra/src/hc_sr04.rs
Comment (brghena) [inline capsules/extra/src/hc_sr04.rs:110]: It might be worth considering timeouts. If you don't receive an EchoStart or EchoEnd interrupt in ~50 ms, something has gone wrong and you can fail the measurement. If you do receive a GPIO interrupt, you can cancel the timer event.

File: capsules/extra/src/hc_sr04.rs
Comment (brghena) [inline capsules/extra/src/hc_sr04.rs:48]: Why do you need an `end_time` member of the struct?

File: capsules/extra/src/hc_sr04.rs
Comment (brghena) [inline capsules/extra/src/hc_sr04.rs:125]: What's guaranteeing that this is in unites of microseconds? Shouldn't it depend on the Alarm Frequency?

**Review by brghena — CHANGES_REQUESTED** [review]
This is great! Pretty simple, but widely used sensor. And this is a pretty darn good implementation. Very clean.

Here are comments for you as I looked through it. Didn't spend much time on the `distance.rs` capsule in this pass.

File: capsules/extra/src/lib.rs
Comment (cristianaprecup) [reply capsules/extra/src/lib.rs:5]: Sorry about that, it was for testing something out. I'll undo the change.

File: kernel/src/hil/sensors.rs
Comment (cristianaprecup) [reply kernel/src/hil/sensors.rs:242]: Thanks, I'll update the comment to be more detailed, similar to the PressureDriver example.

File: capsules/extra/src/hc_sr04.rs
Comment (cristianaprecup) [reply capsules/extra/src/hc_sr04.rs:18]: For 4 meters, the echo time would be about 23 ms under standard temperature and pressure conditions. However, because this applies only at those specific conditions, I used 38 ms to adapt for variations in real-world environments.

File: capsules/extra/src/hc_sr04.rs
Comment (cristianaprecup) [reply capsules/extra/src/hc_sr04.rs:115]: I’ve reduced the pulse duration to 1 ms, and it works well. Thanks for the suggestion.

File: capsules/extra/src/hc_sr04.rs
Comment (cristianaprecup) [reply capsules/extra/src/hc_sr04.rs:48]: I've added a timeout in the `Status::EchoStart` case so that if we don't get an echo within 50ms, the measurement will fail and made sure the timer event gets canceled if we get a GPIO interrupt.
I've also put in a check to make sure the echo pulse duration doesn't exceed the maximum expected time. If it does, we throw an error `ErrorCode::INVAL` since that's likely a bad reading.

File: capsules/extra/src/hc_sr04.rs
Comment (cristianaprecup) [reply capsules/extra/src/hc_sr04.rs:125]: We need the `end_time` member to calculate the duration of the echo pulse. The time difference between `start_time` and `end_time` gives us the travel time of the sound wave, allowing us to compute the distance.

Comment (brghena) [reply capsules/extra/src/hc_sr04.rs]: Given that you only use `end_time` in the same function where it's stored in the struct though, it could just be a local variable rather than a struct member, right?

File: capsules/extra/src/distance.rs
Comment (brghena) [inline capsules/extra/src/distance.rs:30]: Units here would be good.

File: capsules/extra/src/distance.rs
Comment (brghena) [inline capsules/extra/src/distance.rs:126]: If there's an error, don't we also need to clear `self.busy`?

Comment (brghena) [reply capsules/extra/src/distance.rs:126]: Maybe `app.subscribed` should also be set to false too? Otherwise you could later receive unexpected measurement callbacks.

File: capsules/extra/src/distance.rs
Comment (brghena) [inline capsules/extra/src/distance.rs:143]: What if the measurement failed, such as a timeout? Busy will be cleared, but apps won't receive any callback, and their `app.subscribed` will still be set to true.

File: capsules/extra/src/hc_sr04.rs
Comment (brghena) [reply capsules/extra/src/hc_sr04.rs:79]: This got forgotten about.

File: capsules/extra/src/hc_sr04.rs
Comment (brghena) [inline capsules/extra/src/hc_sr04.rs:84]: There actually is a `ticks_from_us()` function that you could use to be more precise. I believe it rounds down though, so you might want `cmp::max(self.alarm.ticks_from_us(1), 1)`.

This is really just an optimization which will speed up measurements though, since 1 ms works fine now. So I don't think it's unreasonable to leave this as-is.

Either way, you should put a comment here documenting why "1" (or "10" if you change it) is a meaningful and correct choice.

File: capsules/extra/src/distance.rs
Comment (brghena) [inline capsules/extra/src/distance.rs:43]: You really ought to make a component for the HC-SR04 distance sensor. https://github.com/tock/tock/tree/master/boards/components

I'm not sure what the best "example" would be here. Maybe the [HD44780 LCD](https://github.com/tock/tock/blob/master/boards/components/src/hd44780.rs) component, which includes GPIO pins and an Alarm, like you do.

**Review by brghena — CHANGES_REQUESTED** [review]
Another round of comments for you.

File: capsules/extra/src/distance.rs
Comment (cristianaprecup) [reply capsules/extra/src/distance.rs:126]: You're right.

File: capsules/extra/src/distance.rs
Comment (cristianaprecup) [reply capsules/extra/src/distance.rs:143]: I’ll add an error callback if there’s a timeout or failure so the app knows and `app.subscribed` gets cleared properly. Thanks

File: capsules/extra/src/hc_sr04.rs
Comment (cristianaprecup) [reply capsules/extra/src/hc_sr04.rs:84]: I've decided to keep it at 1 ms since the measurements are more accurate this way and I’ll add documentation to explain the reasoning.

File: kernel/src/hil/sensors.rs
Comment (bradjc) [inline kernel/src/hil/sensors.rs:247]: I can see where this came from, but I think we should remove `NOSUPPORT`. If the distance sensor can't do this then this trait shouldn't be implemented.

File: capsules/extra/src/hc_sr04.rs
Comment (bradjc) [inline capsules/extra/src/hc_sr04.rs:1]: Please add a `//!` comment with the type of sensor, link to the product, and maybe a quoted description copied from the datasheet if there is a good paragraph.

File: capsules/extra/src/hc_sr04.rs
Comment (bradjc) [inline capsules/extra/src/hc_sr04.rs:33]:
```suggestion
```

This isn't a syscall driver capsule.

File: capsules/extra/src/hc_sr04.rs
Comment (bradjc) [inline capsules/extra/src/hc_sr04.rs:41]: Please move the comments to `///` comments above the enum items so they are included in the docs.

File: capsules/extra/src/hc_sr04.rs
Comment (bradjc) [inline capsules/extra/src/hc_sr04.rs:7]: It's a little confusing what `Client` is when reading the code. I think using the `gpio::` namespace would help since there are both alarm and gpio callbacks.
```suggestion
use kernel::hil::gpio;
```

File: capsules/extra/src/hc_sr04.rs
Comment (bradjc) [inline capsules/extra/src/hc_sr04.rs:50]:
```suggestion
    state: Cell<Status>,
```

The common term for a state machine is `state`, and I think that better matches what this variable to keeping track of.

File: capsules/extra/src/hc_sr04.rs
Comment (bradjc) [inline capsules/extra/src/hc_sr04.rs:138]: This error is not documented in the HIL. `NOACK` probably makes sense but needs to be documented as a possible return value.

File: capsules/extra/src/hc_sr04.rs
Comment (bradjc) [inline capsules/extra/src/hc_sr04.rs:168]: Same here.

File: capsules/extra/src/hc_sr04.rs
Comment (brghena) [inline capsules/extra/src/hc_sr04.rs:153]: Turns out there's a `ticks_to_us` function that would do this for you (and saturate correctly in the case of errors) https://github.com/tock/tock/blob/master/kernel/src/hil/time.rs#L193

File: capsules/extra/src/hc_sr04.rs
Comment (bradjc) [inline capsules/extra/src/hc_sr04.rs:178]: `distance` can be up to 4000 here, correct? For that to be the result, then `duration * SPEED_OF_SOUND` has to be 4000*2000000, but that overflows a u32. Correct me if I have this wrong, but I think this needs to change to avoid overflow.

**Comment by brghena** [comment]
@cristianaprecup there are open comments waiting on you for this PR before we merge it.

There's no rush or anything. People get busy and that's totally okay! But I wanted to make sure you were aware in case you missed them.

If you know you're not going to have time for an extended period, and you want someone else to finish up the PR, let us know. Otherwise we can happily wait a while until you have time to work on this again.

**Comment by cristianaprecup** [comment]
@brghena 
Hello! I have refactored the code to implement the requested changes. However, I encountered an issue with the sensor I'm using for testing, which has caused a slight delay. I'll provide an update later this week once I've resolved the issue and completed proper testing of the implementation. Thank you again for your patience!

**Review by alexandruradovici — CHANGES_REQUESTED** [review]
I just realised that we could actually connect several sensors to this driver. In the current form, there is only one single sensor per driver. I suggest making the following changes:
- allow the system call driver to take an array of sensors `[&dyn Distance]`
- add system call number 1 that returns the number of available sensors
- use one of the `usize` parameters of the `read` system call to select the sensor that you would like a reading from

You can take a loom at how the `Led` driver is implemented.

**Comment by brghena** [comment]
Sorry I haven't responded yet. I'll get to this soon!

File: capsules/extra/src/hc_sr04.rs
Comment (brghena) [inline capsules/extra/src/hc_sr04.rs:165]: Something got messed up here

File: capsules/extra/src/hc_sr04.rs
Comment (brghena) [inline capsules/extra/src/hc_sr04.rs:179]: Something about this line feels wrong to me. `duration` is already in units of microseconds, right? But then you're comparing against units of ticks?

Should this just be `if duration > MAX_ECHO_DELAY_MS * 1000 {`?

File: capsules/extra/src/hc_sr04.rs
Comment (brghena) [inline capsules/extra/src/hc_sr04.rs:194]: To avoid making everything a `u64`, you could restructure this computation. I think it could be `((SPEED_OF_SOUND / 1000) * duration) / (2 * 1000)`

What is the limit on `duration`? If it's 50000 or less, then this will always stay within a 32-bit compuation.

You'd need to add a brief explanation that the equation was rewritten in a less readable way to ensure staying within 32-bit size.

**Review by brghena — CHANGES_REQUESTED** [review]
Some comments after another go through

**Comment by brghena** [comment]
On the point @alexandruradovici made about supporting multiple distance sensors simultaneously.

Personally, I think there's value in this (very well crafted) single distance-sensor driver. So I'd rather have the improvement to supporting multiple distance sensors as a separate PR rather than delaying this one further.

File: doc/syscalls/60009_distance.md
Comment (ppannuto) [inline doc/syscalls/60009_distance.md:9]:
```suggestion
The distance sensor driver allows a process to read the distance measured by a distance sensor. Distance is reported in millimeters.
```

I think this interface looks like it aims to be generic across any distance sensor. If for some reason it should be specific to that exact sensor, it should be in the `0x7____` family of drivers.

**Review by brghena — APPROVED** [review]

File: doc/syscalls/60009_distance.md
Comment (brghena) [reply doc/syscalls/60009_distance.md:9]: I'd agree that the distance interface is generic. So the one-word change here seems appropriate.

**Review by brghena — APPROVED** [review]
This PR looks good-to-go to me!

Thanks so much for all the effort on it.

**Review by alexandruradovici — DISMISSED** [review]

**Comment by alexandruradovici** [comment]
We need to take into account adding a feature to support several distance sensors similar to #4126. I'll merge this, @cristianaprecup please open a new PR when this is implemented.

File: capsules/extra/src/distance.rs
Comment (brghena) [inline capsules/extra/src/distance.rs:54]:
```suggestion
//! ```rust,ignore
```

Need to make tests ignore this like all the other drivers do.

**Review by brghena — APPROVED** [review]

**Pull request merged**
